### PR TITLE
Otsu's Thresholding Augmentation with example

### PR DESCRIPTION
**Pull request by gerasidev**
Otsu's Thresholding Augmentation with example, adding test.

### CODE REVIEW

File: example/src/main/java/de/example/augmentation/OtsuThresholdingAugmentationExample.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+import java.awt.image.BufferedImage;
+import java.io.File;
+import java.io.IOException;
+
+public class OtsuThresholdingAugmentationExample {
+
+    private static final String IMAGE_PATH = "images" + File.separator + "cyborg.png";
+
+    public static void main(String[] args) throws IOException {
+        BufferedImage bufferedImage = loadTestImage(IMAGE_PATH);
```
Comment (Samyssmile) [inline example/src/main/java/de/example/augmentation/OtsuThresholdingAugmentationExample.java:17]: rename to originalImage

File: example/src/main/java/de/example/augmentation/OtsuThresholdingAugmentationExample.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        OtsuThresholdingAugmentation otsuThreshold = new OtsuThresholdingAugmentation();
+        BufferedImage otsoImage = otsuThreshold.apply(bufferedImage);
```
Comment (Samyssmile) [inline example/src/main/java/de/example/augmentation/OtsuThresholdingAugmentationExample.java:19]: otsuImage

File: example/src/main/java/de/example/augmentation/OtsuThresholdingAugmentationExample.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        display(bufferedImage, "Original Image");
+        display(otsoImage, "After Otsu Thresholing");
```
Comment (Samyssmile) [inline example/src/main/java/de/example/augmentation/OtsuThresholdingAugmentationExample.java:21]: "Otsu Augmentation"

File: lib/src/main/java/de/edux/augmentation/effects/OtsuThresholdingAugmentation.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
+package de.edux.augmentation.effects;
+
+import de.edux.augmentation.core.AbstractAugmentation;
+
+import java.awt.*;
+import java.awt.image.BufferedImage;
+
+public class OtsuThresholdingAugmentation extends AbstractAugmentation {
```
Comment (Samyssmile) [inline lib/src/main/java/de/edux/augmentation/effects/OtsuThresholdingAugmentation.java:8]: This class is public API, pls provide a good JavaDoc Documentation at class and public methods.

File: lib/src/main/java/de/edux/augmentation/effects/OtsuThresholdingAugmentation.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+            backgroundWeight += histogram[threshold];
+            if (backgroundWeight == 0) continue;
+            foregroundWeight = totalPixels - backgroundWeight;
+            if (foregroundWeight == 0) break;
+            backgroundPixelIntensity += (float) (threshold * histogram[threshold]);
+
+            float meanBackgroundIntensity = backgroundPixelIntensity / backgroundWeight;
+            float meanForegroundIntensity = (totalPixelIntensity - backgroundPixelIntensity) / foregroundWeight;
+            // Calculate the between-class variance
+            float betweenClassVariance = (float) ((float) backgroundWeight
```
Comment (Samyssmile) [inline lib/src/main/java/de/edux/augmentation/effects/OtsuThresholdingAugmentation.java:73]: remove comment, instead extract the calculation into separate private method called calculateBetweenClassVariance(...)

File: lib/src/main/java/de/edux/augmentation/effects/OtsuThresholdingAugmentation.java [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+                    * (float) foregroundWeight
+                    * Math.pow(meanBackgroundIntensity - meanForegroundIntensity, 2));
+            // Update the maximum variance and optimal
```
Comment (Samyssmile) [inline lib/src/main/java/de/edux/augmentation/effects/OtsuThresholdingAugmentation.java:76]: remove comment

File: lib/src/main/java/de/edux/augmentation/effects/OtsuThresholdingAugmentation.java [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+                int intensity = (pixel >> 16) & 0xff;
+                histogram[intensity]++;
+            }
+        }
+        return histogram;
+    }
+
+    private static int findOtsuThreshold(int[] histogram, int totalPixels) {
+        float totalPixelIntensity = 0;
+        for (int intensity = 0; intensity < 256; intensity++) {
```
Comment (Samyssmile) [inline lib/src/main/java/de/edux/augmentation/effects/OtsuThresholdingAugmentation.java:55]: Replace 256 with a static final field with a propper name

File: lib/src/main/java/de/edux/augmentation/effects/OtsuThresholdingAugmentation.java [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+        BufferedImage grayImage = new BufferedImage(
+                img.getWidth(), img.getHeight(), BufferedImage.TYPE_BYTE_GRAY);
+        Graphics g = grayImage.getGraphics();
+        g.drawImage(img, 0, 0, null);
+        g.dispose();
+        return grayImage;
+    }
+
+    private static int[] calculateHistogram(BufferedImage image) {
+        int[] histogram = new int[256];
```
Comment (Samyssmile) [inline lib/src/main/java/de/edux/augmentation/effects/OtsuThresholdingAugmentation.java:42]: Replace 256 with a static final field with a propper name

File: lib/src/main/java/de/edux/augmentation/effects/OtsuThresholdingAugmentation.java [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+        for (int i = 0; i < image.getWidth(); i++) {
+            for (int j = 0; j < image.getHeight(); j++) {
+                int pixel = image.getRGB(i, j);
+                int intensity = (pixel >> 16) & 0xff;
```
Comment (Samyssmile) [inline lib/src/main/java/de/edux/augmentation/effects/OtsuThresholdingAugmentation.java:46]: Dont use magic numbers. Replace 16 with a static final field with a propper name

Comment (Samyssmile) [reply lib/src/main/java/de/edux/augmentation/effects/OtsuThresholdingAugmentation.java:46]: https://en.wikipedia.org/wiki/Magic_number_(programming)

**Review by Samyssmile — CHANGES_REQUESTED** [review]

File: example/src/main/java/de/example/augmentation/OtsuThresholdingAugmentationExample.java (hunk 2 above)
Comment (gerasidev) [reply example/src/main/java/de/example/augmentation/OtsuThresholdingAugmentationExample.java]: renamed

File: lib/src/main/java/de/edux/augmentation/effects/OtsuThresholdingAugmentation.java (hunk 4 above)
Comment (gerasidev) [reply lib/src/main/java/de/edux/augmentation/effects/OtsuThresholdingAugmentation.java:13]: javaDocs added also refactor methods from static.

File: example/src/main/java/de/example/augmentation/OtsuThresholdingAugmentationExample.java (hunk 3 above)
Comment (gerasidev) [reply example/src/main/java/de/example/augmentation/OtsuThresholdingAugmentationExample.java]: fixed

File: lib/src/main/java/de/edux/augmentation/effects/OtsuThresholdingAugmentation.java (hunk 6 above)
Comment (gerasidev) [reply lib/src/main/java/de/edux/augmentation/effects/OtsuThresholdingAugmentation.java]: removed

File: lib/src/main/java/de/edux/augmentation/effects/OtsuThresholdingAugmentation.java (hunk 7 above)
Comment (gerasidev) [reply lib/src/main/java/de/edux/augmentation/effects/OtsuThresholdingAugmentation.java]: fixed

File: lib/src/main/java/de/edux/augmentation/effects/OtsuThresholdingAugmentation.java (hunk 5 above)
Comment (gerasidev) [reply lib/src/main/java/de/edux/augmentation/effects/OtsuThresholdingAugmentation.java]: fixed

**Review by Samyssmile — APPROVED** [review]

**Pull request merged**